src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -494,69 +505,61 @@ static UCS_CLASS_INIT_FUNC(uct_cuda_copy_iface_t, uct_md_h md, uct_worker_h work
-    mp_params.name            = "CUDA EVENT objects";
-    status = ucs_mpool_init(&mp_params, &self->cuda_event_desc);
-    if (UCS_OK != status) {
-        ucs_error("mpool creation failed");
-        return UCS_ERR_IO_ERROR;
-    }
+    kh_init_inplace(cuda_copy_ctx_rscs, &self->ctx_rscs);
 
     ucs_queue_head_init(&self->active_queue);
 
+    return UCS_OK;
+}
+
+static void uct_cuda_copy_stream_destroy(CUstream *stream, int valid_ctx)
+{
+    if ((*stream == NULL) || (!valid_ctx)) {
+        return;
+    }
+
+    UCT_CUDADRV_FUNC_LOG_WARN(cuStreamDestroy(*stream));
+    *stream = NULL;
+}
+
+static void uct_cuda_copy_ctx_rsc_destroy(uct_cuda_copy_ctx_rsc_t *ctx_rsc)
+{
+    int ctx_rsc_valid = uct_cuda_copy_ctx_rsc_valid(ctx_rsc);
+    ucs_memory_type_t src, dst;
+    ucs_queue_head_t *event_q;
+
     ucs_memory_type_for_each(src) {
         ucs_memory_type_for_each(dst) {
-            self->queue_desc[src][dst].stream = 0;
-            ucs_queue_head_init(&self->queue_desc[src][dst].event_queue);
+            event_q = &ctx_rsc->queue_desc[src][dst].event_queue;
+            if (!ucs_queue_is_empty(event_q)) {
+                ucs_warn("stream destroyed but queue not empty");
+            }
+
+            uct_cuda_copy_stream_destroy(&ctx_rsc->queue_desc[src][dst].stream,
+                                         ctx_rsc_valid);
         }
     }
 
-    self->short_stream = 0;
-    self->cuda_context = 0;
-
-    return UCS_OK;
+    uct_cuda_copy_stream_destroy(&ctx_rsc->short_stream, ctx_rsc_valid);
+    ucs_mpool_cleanup(&ctx_rsc->cuda_event_desc, 1);
 }
 
 static UCS_CLASS_CLEANUP_FUNC(uct_cuda_copy_iface_t)
 {
-    CUstream *stream;
-    CUcontext cuda_context;
-    ucs_queue_head_t *event_q;
-    ucs_memory_type_t src, dst;
+    unsigned long long ctx_id;
+    khiter_t iter;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
 
     uct_base_iface_progress_disable(&self->super.super.super,
                                     UCT_PROGRESS_SEND | UCT_PROGRESS_RECV);
 
-    UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetCurrent(&cuda_context));
-    if (uct_cuda_base_context_match(cuda_context, self->cuda_context)) {
-
-        ucs_memory_type_for_each(src) {
-            ucs_memory_type_for_each(dst) {
-                stream  = &self->queue_desc[src][dst].stream;
-                event_q = &self->queue_desc[src][dst].event_queue;
-
-                if (!ucs_queue_is_empty(event_q)) {
-                    ucs_warn("stream destroyed but queue not empty");
-                }
-
-                if (*stream == 0) {
-                    continue;
-                }
-
-                UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(*stream));
-            }
-        }
-
-        if (self->short_stream) {
-            UCT_CUDADRV_FUNC_LOG_ERR(cuStreamDestroy(self->short_stream));
-        }
-    }
+    kh_foreach_key(&self->ctx_rscs, ctx_id, {
+        iter    = kh_get(cuda_copy_ctx_rscs, &self->ctx_rscs, ctx_id);
+        ctx_rsc = &kh_value(&self->ctx_rscs, iter);
```

Comment:
given the queue head/hash issue, maybe the hash table should point to ctx_rsc and not store the struct directly. then we can use kh_foreach_value here

Reply:
Updated hash map value type.